### DISCUSSION TITLE
remake the api reference?

### DISCUSSION THREAD

**Discussion by TrainzMarcel** [gilzoide/godot-lua-pluginscript#30] [Ideas]
i cant find anything about finding screen size in them, yet by trial and error i found u can use OS.window_size to get a vector2 of the window size. im assuming this is true with a lot of other things. at the same time, im not sure how to warp the mouse. i also struggled to find out how to access other nodes (i had to find the discussion in here) as the reference doesnt make it very easy to find things. if this idea is seriously considered, ill help however i can to make a better reference

it would also be great if there was a godot lua discord server

**Comment by TrainzMarcel** [comment]
i apologize, i figured it out

**Reply by gilzoide** [reply]
Sorry about the delay in answering, this week I was really busy. I'm glad you sorted it out.
Feel free to ask again if you need any more help ^^

**Comment by gilzoide** [comment]
Hey @TrainzMarcel
To be honest, I don't think remaking the reference API is necessary.
The Godot API is dynamic and 100% accessible at runtime without static bindings, so you can access all of it from Lua just like you would in GDScript. This is nice because new APIs from new versions of Godot are available to the plugin without any changes.
The only difference is the language used (e.g. when calling methods in Lua, you need to use a colon, like `OS:alert("hi there!")`).

So if you want to know how to do something Godot-related with Lua PluginScript, you find out how to do it in GDScript, then translate this code to Lua.

Now, I agree the documentation is lacking a lot, like explaining that you can actually access all the Godot API automatically, have a "From GDScript to Lua" doc explaining how to do things in Lua for people who know GDScript, these kinds of things. I'll make an effort to work on this GDScript to Lua guide this weekend, I'll let you know.

**Reply by TrainzMarcel** [reply]
hello gilzoide,
i am extremely greatful for the lua binding, i got familiar with it and programming my game with it is a lot of fun and i save a lot of time and i really cant thank u enough for it. i have one last question though, is there a way to directly set member variables of objects? say i have an icon and i type self.position.x = 500. this doesnt throw an error but it doesnt do anything, im only able to read the value, and unable to directly write to it. the set functions from the gdscript reference are also nil values in the table.

**Reply by gilzoide** [reply]
> programming my game with it is a lot of fun and i save a lot of time and i really cant thank u enough for it

I'm glad to hear it ^^
No problem, if you need anything, just give it a shout!

> say i have an icon and i type self.position.x = 500. this doesnt throw an error but it doesnt do anything

Oh, I know that one.
The thing in this particular case is that `self.position` returns a brand new `Vector2` value.
When you run `self.position.x = 500`, you're actually changing the X value of this new `Vector2` that gets returned, instead of setting the nested value of `position.x` in `self`.
What GDScript does behind the scenes, that you can do manually, is make an [indexed set](https://docs.godotengine.org/en/stable/classes/class_object.html#class-object-method-set-indexed) to the `position:x` property path:
```lua
self:set_indexed('position:x', 500)
```

Alternatively, you can always get the current `position`, modify it, then set it fully again:
```lua
local pos = self.position
pos.x = 500
self.position = pos
```
Or even create the new `Vector2` in a single line using the previous Y value:
```lua
self.position = Vector2(500, self.position.y)
```

**Reply by TrainzMarcel** [reply]
thanks, i got it

**Comment by gilzoide** [comment]
As for a Godot Lua Discord Server, it could be cool, yeah. I don't use Discord much and prefer a persistent forum-like channel like this one at GitHub, but we could create one if people are willing to use it, no problem!

**Reply by TrainzMarcel** [reply]
in that case, im going to make one in a week or two :3

**Reply by TrainzMarcel** [reply]
i made it, heres the invite https://discord.gg/rZJbXZXQWe

**Reply by gilzoide** [reply]
Nice! Feel free to open a PR adding a badge in the README, probably with a link like https://img.shields.io/discord/1056941025559588874.svg (reference: https://github.com/Naereen/badges#discord)
